src/client/datascience/interactive-window/nativeInteractiveWindow.ts

```diff
@@ -130,21 +135,30 @@ export class NativeInteractiveWindow implements IInteractiveWindowLoadable {
             this._submitters.push(owner);
         }
 
+        this.initialControllerSelected = createDeferred<void>();
+
+        // Immediately try to find a selected controller for our NotebookDocument,
+        // as it's possible that a selection event fired before our ctor was able to run
+        const controller = this.notebookControllerManager.getSelectedNotebookController(this.notebookDocument);
```

Comment:
@joyceerhl is this necessary? I  can't think of a situation when `controller` here would ever be not undefined. for a controller to be associaed with an interactive window, the user has to run a cell, but to run a cell the window has to appear and user has to select a controller.
Personally i'd prefer to get rid of `getSelecedNotebookController`, i thought i removed this in the main branch (I've added this back in my other repo).

Reply:
The problem is that the selection event can fire before the interactive window ctor runs. It's either we maintain the mapping in the notebookControllerManager, or we maintain the same mapping in the nativeInteractiveWindowProvider and have the provider set the controller on the nativeInteractiveWindow.

> to run a cell the window has to appear and user has to select a controller

Yes, but the controller may not have been selected by the time that the interactive window's addCode method gets called.

Basically what we want to do is ensure that execution requests are not dropped even when a controller hasn't been selected yet, which can absolutely happen. Personally it makes more sense to me to wait for the controller to be selected in the interactiveWindow, versus exposing that information say in the commandListener and making the commandListener responsible for checking if a controller has been selected and queuing execution requests until the controller has been selected.

Reply:
Hmm, lets chat tomorrow.